reps/2023-07-28-serve-grpc-proxy.md

```diff
@@ -0,0 +1,633 @@
+### Required Reviewers
+@edoakes, @ericl, @pcmoritz
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@edoakes
+
+## Design and Architecture
+### High Level Description
+- We will serve gRPC from the same actors that currently serve HTTP
+- User will create their own protobufs, services, and methods
+- User will start Serve's gRPC proxy with port and their protobuf servicer functions
+  (servicer functions are autogenerated by `.proto` file using grpc tool that defines
+  how the gRPC server should handle the requests)
+- User will call gRPC services using user defined protobufs, and with optional metadata
+- Client will receive user defined protobuf as response
+- Serve will also return request id back to the client with trailing metadata
+- gRPC deployments will have full feature parity with HTTP, including
+  streaming responses, model multiplexing, and model composition support
+
+
+### Example code
+#### Serve Config gRPC Option
+```python
+class gRPCOptionsSchema(BaseModel, extra=Extra.forbid):
+  """Options to start the gRPC Proxy with."""
+
+  port: int = Field(
+    default=-1,
+    description=(
+      "Port for gRPC server. Will only start gRPC server if set. Cannot be "
+      "updated once Serve has started running. Serve must be shut down and "
+      "restarted with the new port instead."
+    ),
+  )
+  grpc_servicer_functions: List[str] = Field(
+    default=[],
+    description=(
+      "The servicer functions used to add the method handlers to the gRPC server."
+      "Default to empty list, which means no gRPC methods will be added"
+      "and no gRPC server will be started. The servicer functions need to be"
+      "importable from the context of where Serve is running."
+    ),
+  )
+
+```
+
+#### Accepted Metadata
+- `application`: The name of the Serve application to route to. If not passed, first
+  application will be used automatically.
+- `request_id`: The request id to track the request. Feature parity with http proxy.
+- `multiplexed_model_id`: The model id to do model multiplexing. Feature parity with
+  http proxy.
+
+#### Unary gRPC example
+```proto
+// user_defined_protos.proto
+
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "io.ray.examples.user_defined_protos";
+option java_outer_classname = "UserDefinedProtos";
+
+package userdefinedprotos;
+
+message UserDefinedMessage {
+  string name = 1;
+  string foo = 2;
+  int64 num = 3;
+}
+
+message UserDefinedResponse {
+  string greeting = 1;
+  int64 num_x2 = 2;
+}
+
+message UserDefinedMessage2 {}
+
+message UserDefinedResponse2 {
+  string greeting = 1;
+}
+
+service UserDefinedService {
+  rpc __call__(UserDefinedMessage) returns (UserDefinedResponse);
+  rpc Method1(UserDefinedMessage) returns (UserDefinedResponse);
+  rpc Method2(UserDefinedMessage2) returns (UserDefinedResponse2);
+  rpc Streaming(UserDefinedMessage) returns (stream UserDefinedResponse);
+}
+
+```
+
+```yaml
+# config.yaml
+
+grpc_options:
```

Comment:
To confirm: Will config generation support be added to the `serve build` CLI?

Reply:
Yes!